staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go

```diff
@@ -170,17 +171,20 @@ func (c *codec) Encode(obj runtime.Object, w io.Writer) error {
 	case *runtime.Unknown:
 		return c.encoder.Encode(obj, w)
 	case runtime.Unstructured:
-		// avoid conversion roundtrip if GVK is the right one already or is empty (yes, this is a hack, but the old behaviour we rely on in kubectl)
-		objGVK := obj.GetObjectKind().GroupVersionKind()
-		if len(objGVK.Version) == 0 {
-			return c.encoder.Encode(obj, w)
-		}
-		targetGVK, ok := c.encodeVersion.KindForGroupVersionKinds([]schema.GroupVersionKind{objGVK})
-		if !ok {
-			return runtime.NewNotRegisteredGVKErrForTarget(objGVK, c.encodeVersion)
-		}
-		if targetGVK == objGVK {
-			return c.encoder.Encode(obj, w)
+		// Only avoid conversion if unstructured is not a list, giving the converter a chance to convert list items.
```

Comment:
@sttts @mbohlool it's hard to know what to make of this. any words of wisdom for someone trying to imagine the scenario that triggers this?

Reply:
This was the problem this part of the code tries to fix:

Assume v1 is the storage version and v2 is the request version. When we request backend to list objects in v2 format, it return a v2 list with items from the storage. But those items are in v1 (or some of them are at least). The we would have a list that its own version is v2 but its items may have other versions. This part of the code were treating UnstructuredList as an Unstructured object and because requested and target version are the same (v2) it won't call our converters on it. By adding this code, we let the converter to have a chance to look at the items in a list even the list GVK itself is the same as target.

Reply:
Prefering the explicit type very much. We should base this logic on kinds (= Golang types for native types), not on the shape of the object. So 👍  for the comment.